### PR TITLE
Improve transactional integrity for starting controller jobs in dispatcherd

### PR DESCRIPTION
**Pull request by fosterseth**
## Summary
- Remove `SELECT FOR UPDATE` / `transaction.atomic()` from `BaseTask.transition_status` which was causing increased database transaction rollbacks observed in perfscale testing
- Move status transition into `dispatch_waiting_jobs`, using a conditional `filter().update()` that is atomic at the DB level without explicit row locks
- Delete the `transition_status` method entirely — it was an artifact of the feature flag era and unnecessary given `dispatch_waiting_jobs` is already a singleton (`on_duplicate='queue_one'`) scoped to the local node
- Status is now updated after task submission to dispatcherd, so the job's UUID is in the dispatch pipeline before being marked `running` — this prevents the reaper from incorrectly reaping jobs during the handoff window
- `BaseTask.run` handles the race where a worker picks up the task before the status update lands by accepting `waiting` and transitioning it to `running` itself

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Touches core job start/cancel state transitions and could affect edge-case concurrency (e.g., fast workers or reaping timing), though changes are localized and covered by new tests.
> 
> **Overview**
> **Job start status handling is refactored to reduce DB locking and tighten the dispatch-to-execution handoff.** `dispatch_waiting_jobs` now conditionally updates `UnifiedJob` rows from `waiting` to `running` immediately after submitting the task to dispatcherd.
> 
> `BaseTask.run` removes the `transaction.atomic()`/`select_for_update()`-based `transition_status` logic (and deletes the method), and instead tolerates `waiting` on worker pickup by flipping it to `running` before continuing; it also explicitly cancels early if `cancel_flag` is set. Adds a functional test asserting a `waiting` job is `running` by the time `pre_run_hook` executes.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 72f291b9eef0c0060ab4f2491b8c7ae83988f891. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

## Summary by CodeRabbit

* **Refactor**
  * Streamlined job dispatch and execution status handling with unified status transition logic.

* **Breaking Changes**
  * Removed a public method from the job task API that handled manual status transitions.

### CODE REVIEW

**Comment by fosterseth** [comment]
@AlanCoding 
a few open questions from me:

- [ ] this does change the meaning of "running" in our system. Before running meant the dispatcher worker is actually handling that job. This change means "running" is right before we submit task to parent dispatcher process. I do wonder if this is going to make debugging stuck jobs in the future more difficult
- [ ] do we need a select for update / row locking? I am guessing "no" since we are in a code block that _should_ only be atomically running locally (no duplicates)
- [ ] we still need to handle cancel flag right?

**Comment by AlanCoding** [comment]
> do we need a select for update / row locking? I am guessing "no" since we are in a code block that should only be atomically running locally (no duplicates)

Yes, I believe that would be true.

> this does change the meaning of "running" in our system. Before running meant the dispatcher worker is actually handling that job. This change means "running" is right before we submit task to parent dispatcher process. I do wonder if this is going to make debugging stuck jobs in the future more difficult

There's a form of accounting here to worry about. Consider the topic of "reaping" jobs in the running status. This probably needs some updating to account for the special state where the dispatcher main process is still running looking for a worker to hand the RunJob.run task to, or something like that. To make this complete and trackable, I believe that the task submission to the `binder` needs to happen _before_ the status is changed (it is not now). This should leave the job in the local task queue, even if it is blocked on needing an available worker (if this was to be iron-clad, we would want to do a `control_with_reply` but I currently think this is overkill). And the reper also needs to consider jobs that are queued for reaping running jobs.

> we still need to handle cancel flag right?

I would say "no". Because the best thing we can do for a canceled job is to submit RunJob.run, and let that handle it.

File: awx/main/tasks/jobs.py [hunk 1]
```diff
@@ -551,44 +551,19 @@ def final_run_hook(self, instance, status, private_data_dir):
...
-        if not self.instance:  # Used to skip fetch for local runs
-            if not self.transition_status(pk):
-                logger.info(f'Job {pk} is being ran by another process, exiting')
-                return
-
-        # Load the instance
+        # Load the instance - status should be "running" from dispatch_waiting_jobs,
+        # but may still be "waiting" if the worker picked this up before the status update landed.
         self.instance = self.update_model(pk)
+        if self.instance.status == 'waiting':
```
Comment (fosterseth) [inline awx/main/tasks/jobs.py:567]: @AlanCoding 
do you have a proposed alternative here?

File: awx/main/tasks/jobs.py [hunk 2]
```diff
@@ -551,44 +551,17 @@ def final_run_hook(self, instance, status, private_data_dir):
...
-            # .values - avoid loading a full object, this is known to lead to deadlocks due to signals
-            #   the signals load other related rows which another process may be locking, and happens in practice
-            # of=('self',) - keeps FK tables out of the lock list, another way deadlocks can happen
-            # .get - just load the single job
-            instance_data = UnifiedJob.objects.select_for_update(of=('self',)).values('status', 'cancel_flag').get(pk=pk)
-
-            # If status is not waiting (obtained under lock) then this process does not have clearence to run
-            if instance_data['status'] == 'waiting':
-                if instance_data['cancel_flag']:
-                    updated_status = 'canceled'
```
[the remark this answers is not in the document]
Comment (fosterseth) [reply awx/main/tasks/jobs.py:567]: we're aware of this, run() will detect cancelled jobs and handle it

File: awx/main/tasks/jobs.py [hunk 3]
```diff
@@ -205,6 +204,7 @@ def dispatch_waiting_jobs(binder):
         if not kwargs:
             kwargs = {}
         binder.control('run', data={'task': serialize_task(uj._get_task_class()), 'args': [uj.id], 'kwargs': kwargs, 'uuid': uj.celery_task_id})
+        UnifiedJob.objects.filter(pk=uj.pk, status='waiting').update(status='running', start_args='')
```
[the remark this answers is not in the document]
Comment (fosterseth) [reply awx/main/tasks/jobs.py:207]: this is the only place in code where we transition to from waiting to running, and only one instance of this method can be running at a given time due to on_duplication rule

File: awx/main/tasks/jobs.py [hunk 4]
```diff
@@ -200,10 +199,19 @@ def _wrapped(self, *args, **kwargs):
...
+            UnifiedJob.objects.filter(pk=uj.pk, status='waiting').update(status='canceled', start_args='')
+            continue
         kwargs = uj.get_start_kwargs()
         if not kwargs:
             kwargs = {}
+        updated = UnifiedJob.objects.filter(pk=uj.pk, status='waiting', cancel_flag=False).update(status='running', start_args='')
+        if not updated:
+            # Status or cancel_flag changed since our read, skip this job
+            continue
         binder.control('run', data={'task': serialize_task(uj._get_task_class()), 'args': [uj.id], 'kwargs': kwargs, 'uuid': uj.celery_task_id})
```
[the remark this answers is not in the document]
Comment (fosterseth) [reply awx/main/tasks/jobs.py]: i reordered it

File: awx/main/tasks/jobs.py [hunk 5]
```diff
@@ -551,44 +551,23 @@ def final_run_hook(self, instance, status, private_data_dir):
...
-            if not self.transition_status(pk):
-                logger.info(f'Job {pk} is being ran by another process, exiting')
-                return
-
-        # Load the instance
+        # Load the instance - status should be "running" from dispatch_waiting_jobs,
+        # but may still be "waiting" if the worker picked this up before the status update landed.
         self.instance = self.update_model(pk)
+        if self.instance.status == 'waiting':
+            if self.instance.cancel_flag:
```
Comment (fosterseth) [inline awx/main/tasks/jobs.py]: `test_cancel_flag_on_start` was failing, we need to check cancel flag somewhere

File: awx/main/tasks/jobs.py (hunk 1 above)
Comment (AlanCoding) [reply awx/main/tasks/jobs.py:567]: I don't know entirely what you're trying to address here. The cancelled_flag check is needed, but I don't think that should be conditional on the waiting status as it is here.

Honestly I don't think we should ever go into this in waiting status? Couldn't we let the later `if self.instance.status != 'running':` error that case? Is there some test case this needs to satisfy that we can't change for some reason?

File: awx/main/tasks/jobs.py [hunk 6]
```diff
@@ -551,44 +551,23 @@ def final_run_hook(self, instance, status, private_data_dir):
...
         """
-        if not self.instance:  # Used to skip fetch for local runs
-            if not self.transition_status(pk):
-                logger.info(f'Job {pk} is being ran by another process, exiting')
-                return
-
-        # Load the instance
+        # Load the instance - status should be "running" from dispatch_waiting_jobs,
+        # but may still be "waiting" if the worker picked this up before the status update landed.
         self.instance = self.update_model(pk)
```
Comment (AlanCoding) [inline awx/main/tasks/jobs.py]: I made it so the local project update will have the instance passed locally, and in that case `self.instance` will be non-null, so I would like to keep that non-null check, so we can do nothing in that case.

File: awx/main/tasks/jobs.py (hunk 5 above)
Comment (AlanCoding) [reply awx/main/tasks/jobs.py]: yeah, it's a really good idea to check the cancel flag! At whatever point `self.instance` is defined, and preferably after the other status check (expecting running status), to prefer that error status over switching status when it's already non-running, if that makes sense.

File: awx/main/tasks/jobs.py (hunk 1 above)
Comment (fosterseth) [reply awx/main/tasks/jobs.py:567]: it's to prevent the race condition -- dispatch_waiting_jobs submits the task to dispatcherd before changing to Running. Therefore, it is possible that the job is still in waiting by the time RunJob.run() runs

Comment (AlanCoding) [reply awx/main/tasks/jobs.py:567]: ok, then I think it would be pretty safe to allow status in [running, waiting]

File: awx/main/tasks/jobs.py (hunk 5 above)
Comment (fosterseth) [reply awx/main/tasks/jobs.py]: i'm now checking for cancel_flag after checking the running status

File: awx/main/tasks/jobs.py (hunk 6 above)
Comment (fosterseth) [reply awx/main/tasks/jobs.py]: fixed

File: awx/main/tasks/jobs.py [hunk 7]
```diff
@@ -551,47 +551,28 @@ def final_run_hook(self, instance, status, private_data_dir):
...
         Run the job/task and capture its output.
         """
         if not self.instance:  # Used to skip fetch for local runs
-            if not self.transition_status(pk):
-                logger.info(f'Job {pk} is being ran by another process, exiting')
+            # Load the instance - status should be "running" from dispatch_waiting_jobs,
+            # but may still be "waiting" if the worker picked this up before the status update landed.
+            self.instance = self.update_model(pk)
+
+            if self.instance.status == 'waiting':
```
Comment (AlanCoding) [inline awx/main/tasks/jobs.py]: I would still de-indent this? To get out from under the `if not self.instance:` conditional.

Comment (fosterseth) [reply awx/main/tasks/jobs.py]: addressed this

**Review by AlanCoding — APPROVED** [review]

File: awx/main/tests/functional/tasks/test_tasks_jobs.py [hunk 8]
```diff
@@ -29,3 +29,24 @@ def test_cancel_flag_on_start(jt_linked, caplog):
...
+        instance.refresh_from_db()
+        assert instance.status == 'running'
+
+    mocker.patch.object(RunJob, 'pre_run_hook', side_effect=check_status_is_running)
+
+    task = RunJob()
+    try:
+        task.run(job.id)
+    except Exception:
+        pass
```
[the remark this answers is not in the document]
Comment (fosterseth) [reply awx/main/tests/functional/tasks/test_tasks_jobs.py:55]: good point, I've addressed this
